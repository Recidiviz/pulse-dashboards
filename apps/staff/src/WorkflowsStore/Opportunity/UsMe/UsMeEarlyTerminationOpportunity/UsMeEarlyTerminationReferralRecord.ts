// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { z } from "zod";

import { dateStringSchema, opportunitySchemaBase } from "~datatypes";

import { NullCoalesce } from "../../schemaHelpers";

const usMePaidAllOwedRestitution = NullCoalesce(
  {},
  z.object({ amountOwed: z.number().optional() }),
).optional();

const eligibleCriteria = z.object({
  usMePaidAllOwedRestitution,
  noConvictionWithin6Months: NullCoalesce(
    {},
    z
      .object({
        latestConvictions: z.array(z.string()).optional(),
      })
      .optional(),
  ).optional(),
  usMeSupervisionPastHalfFullTermReleaseDateFromProbationStart: z.object({
    eligibleDate: dateStringSchema,
  }),
  usMeNoPendingViolationsWhileSupervised: NullCoalesce(
    {},
    z
      .object({
        currentStatus: z.string().optional(),
        violationDate: dateStringSchema.optional(),
      })
      .optional(),
  ).optional(),
  supervisionLevelIsMediumOrLower: z.object({
    supervisionLevel: z.string(),
  }),
});

const ineligibleCriteria = z.object({
  usMePaidAllOwedRestitution,
  usMeNoPendingViolationsWhileSupervised: z
    .object({
      currentStatus: z.string(),
      violationDate: dateStringSchema,
    })
    .nullable()
    .optional(),
});

export const usMeEarlyTerminationSchema = opportunitySchemaBase.extend({
  eligibleCriteria,
  ineligibleCriteria,
});

export type UsMeEarlyTerminationReferralRecord = z.infer<
  typeof usMeEarlyTerminationSchema
>;

export type UsMeEarlyTerminationReferralRecordRaw = z.input<
  typeof usMeEarlyTerminationSchema
>;
