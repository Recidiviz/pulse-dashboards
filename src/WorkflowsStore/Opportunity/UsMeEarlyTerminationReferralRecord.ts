// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { z } from "zod";

import { TransformFunction } from "../subscriptions";
import {
  caseNotesSchema,
  dateStringSchema,
  NullCoalesce,
  opportunitySchemaBase,
} from "./schemaHelpers";

const criteria = z.object({
  usMePaidAllOwedRestitution: NullCoalesce(
    {},
    z.object({ amountOwed: z.number().optional() })
  ).optional(),
  noConvictionWithin6Months: NullCoalesce(
    {},
    z
      .object({
        latestConvictions: z.array(z.string()).optional(),
      })
      .optional()
  ).optional(),
  usMeSupervisionPastHalfFullTermReleaseDateFromProbationStart: NullCoalesce(
    {},
    z.object({
      eligibleDate: dateStringSchema.optional(),
    })
  ).optional(),
  onMediumSupervisionLevelOrLower: z.object({ supervisionLevel: z.string() }),
  usMeNoPendingViolationsWhileSupervised: NullCoalesce(
    {},
    z
      .object({
        currentStatus: z.string().optional(),
        violationDate: dateStringSchema.optional(),
      })
      .optional()
  ).optional(),
});

const ineligibleCriteria = z.object({
  usMeNoPendingViolationsWhileSupervised: z
    .object({
      currentStatus: z.string(),
      violationDate: dateStringSchema,
    })
    .nullable()
    .optional(),
});

export const usMeEarlyTerminationSchema = opportunitySchemaBase
  .extend({
    eligibleCriteria: criteria,
    ineligibleCriteria: criteria
      .pick({
        usMePaidAllOwedRestitution: true,
        usMeSupervisionPastHalfFullTermReleaseDateFromProbationStart: true,
      })
      .merge(ineligibleCriteria),
  })
  .merge(caseNotesSchema);

export type UsMeEarlyTerminationReferralRecord = z.infer<
  typeof usMeEarlyTerminationSchema
>;

export type UsMeEarlyTerminationReferralRecordRaw = z.input<
  typeof usMeEarlyTerminationSchema
>;

export const transformReferral: TransformFunction<
  UsMeEarlyTerminationReferralRecord
> = (record) => {
  return usMeEarlyTerminationSchema.parse(record);
};
