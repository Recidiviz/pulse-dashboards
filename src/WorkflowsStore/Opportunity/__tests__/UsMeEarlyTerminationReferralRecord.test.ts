// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { transformReferral } from "../UsMeEarlyTerminationReferralRecord";

test("transform record with restitution case", () => {
  const rawRecord = {
    stateCode: "US_ME",
    externalId: "abc123",
    eligibleCriteria: {
      usMePaidAllOwedRestitution: { amountOwed: 0 },
      noConvictionWithin6Months: null,
      onMediumSupervisionLevelOrLower: {
        supervisionLevel: "MEDIUM",
      },
      supervisionPastHalfFullTermReleaseDateFromSupervisionStart: {
        eligibleDate: "2022-01-03",
      },
    },
    ineligibleCriteria: {},
  };

  expect(transformReferral(rawRecord)).toMatchSnapshot();
});

test("transform record without restitution case", () => {
  const rawRecord = {
    stateCode: "US_ME",
    externalId: "abc123",
    eligibleCriteria: {
      usMePaidAllOwedRestitution: null,
      noConvictionWithin6Months: null,
      onMediumSupervisionLevelOrLower: {
        supervisionLevel: "MEDIUM",
      },
      supervisionPastHalfFullTermReleaseDateFromSupervisionStart: {
        eligibleDate: "2022-01-03",
      },
    },
    ineligibleCriteria: {},
  };

  expect(transformReferral(rawRecord)).toMatchSnapshot();
});

test("transform record with restitution ineligibleCriteria", () => {
  const rawRecord = {
    stateCode: "US_ME",
    externalId: "abc123",
    eligibleCriteria: {
      noConvictionWithin6Months: null,
      onMediumSupervisionLevelOrLower: {
        supervisionLevel: "MEDIUM",
      },
      supervisionPastHalfFullTermReleaseDateFromSupervisionStart: {
        eligibleDate: "2022-01-03",
      },
    },
    ineligibleCriteria: {
      usMePaidAllOwedRestitution: {
        amountOwed: 500,
      },
    },
  };

  expect(transformReferral(rawRecord)).toMatchSnapshot();
});

test("transform record with pending violation ineligibleCriteria", () => {
  const rawRecord = {
    stateCode: "US_ME",
    externalId: "abc123",
    eligibleCriteria: {
      noConvictionWithin6Months: null,
      onMediumSupervisionLevelOrLower: {
        supervisionLevel: "MEDIUM",
      },
      supervisionPastHalfFullTermReleaseDateFromSupervisionStart: {
        eligibleDate: "2022-01-03",
      },
    },
    ineligibleCriteria: {
      usMeNoPendingViolationsWhileSupervised: {
        currentStatus: "Pending Violation",
        violationDate: "2022-07-13",
      },
    },
  };

  expect(transformReferral(rawRecord)).toMatchSnapshot();
});

test("transform record with null pending violation criteria", () => {
  const rawRecord = {
    stateCode: "US_ME",
    externalId: "abc123",
    eligibleCriteria: {
      noConvictionWithin6Months: null,
      onMediumSupervisionLevelOrLower: {
        supervisionLevel: "MEDIUM",
      },
      supervisionPastHalfFullTermReleaseDateFromSupervisionStart: {
        eligibleDate: "2022-01-03",
      },
      usMeNoPendingViolationsWhileSupervised: null,
    },
    ineligibleCriteria: {},
  };

  expect(transformReferral(rawRecord)).toMatchSnapshot();
});
