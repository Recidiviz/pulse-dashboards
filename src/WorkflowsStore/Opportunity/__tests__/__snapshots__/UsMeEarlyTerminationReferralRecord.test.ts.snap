// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform record with null pending violation criteria 1`] = `
Object {
  "caseNotes": Object {},
  "eligibleCriteria": Object {
    "noConvictionWithin6Months": Object {},
    "onMediumSupervisionLevelOrLower": Object {
      "supervisionLevel": "MEDIUM",
    },
    "supervisionPastHalfFullTermReleaseDateFromSupervisionStart": Object {
      "eligibleDate": 2022-01-03T00:00:00.000Z,
    },
    "usMeNoPendingViolationsWhileSupervised": Object {},
  },
  "externalId": "abc123",
  "ineligibleCriteria": Object {},
  "stateCode": "US_ME",
}
`;

exports[`transform record with pending violation ineligibleCriteria 1`] = `
Object {
  "caseNotes": Object {},
  "eligibleCriteria": Object {
    "noConvictionWithin6Months": Object {},
    "onMediumSupervisionLevelOrLower": Object {
      "supervisionLevel": "MEDIUM",
    },
    "supervisionPastHalfFullTermReleaseDateFromSupervisionStart": Object {
      "eligibleDate": 2022-01-03T00:00:00.000Z,
    },
  },
  "externalId": "abc123",
  "ineligibleCriteria": Object {
    "usMeNoPendingViolationsWhileSupervised": Object {
      "currentStatus": "Pending Violation",
      "violationDate": 2022-07-13T00:00:00.000Z,
    },
  },
  "stateCode": "US_ME",
}
`;

exports[`transform record with restitution case 1`] = `
Object {
  "caseNotes": Object {},
  "eligibleCriteria": Object {
    "noConvictionWithin6Months": Object {},
    "onMediumSupervisionLevelOrLower": Object {
      "supervisionLevel": "MEDIUM",
    },
    "supervisionPastHalfFullTermReleaseDateFromSupervisionStart": Object {
      "eligibleDate": 2022-01-03T00:00:00.000Z,
    },
    "usMePaidAllOwedRestitution": Object {
      "amountOwed": 0,
    },
  },
  "externalId": "abc123",
  "ineligibleCriteria": Object {},
  "stateCode": "US_ME",
}
`;

exports[`transform record with restitution ineligibleCriteria 1`] = `
Object {
  "caseNotes": Object {},
  "eligibleCriteria": Object {
    "noConvictionWithin6Months": Object {},
    "onMediumSupervisionLevelOrLower": Object {
      "supervisionLevel": "MEDIUM",
    },
    "supervisionPastHalfFullTermReleaseDateFromSupervisionStart": Object {
      "eligibleDate": 2022-01-03T00:00:00.000Z,
    },
  },
  "externalId": "abc123",
  "ineligibleCriteria": Object {
    "usMePaidAllOwedRestitution": Object {
      "amountOwed": 500,
    },
  },
  "stateCode": "US_ME",
}
`;

exports[`transform record without restitution case 1`] = `
Object {
  "caseNotes": Object {},
  "eligibleCriteria": Object {
    "noConvictionWithin6Months": Object {},
    "onMediumSupervisionLevelOrLower": Object {
      "supervisionLevel": "MEDIUM",
    },
    "supervisionPastHalfFullTermReleaseDateFromSupervisionStart": Object {
      "eligibleDate": 2022-01-03T00:00:00.000Z,
    },
    "usMePaidAllOwedRestitution": Object {},
  },
  "externalId": "abc123",
  "ineligibleCriteria": Object {},
  "stateCode": "US_ME",
}
`;
